### PR TITLE
Added easy_profiler 2.1.0

### PR DESCRIPTION
**Pull request by GavinNL**
**easy_profiler/2.1.0**

Tested on Linux Mint 20 using gcc9 and clang 9

easy_profiler is a profiling tool with a user interface. Unfortunately the user interface is built using Qt so it is disabled during building. 

The compile definition `BUILD_WITH_EASY_PROFILER` should be set for any project that plans to enable the profiling. This definition is NOT added to the conan definitions because profiling can be compile-time disabled even when the library is linked

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/easy_profiler/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 2.8.11)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} example.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (uilianries) [inline recipes/easy_profiler/all/test_package/CMakeLists.txt:8]:
```suggestion
target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
```

File: recipes/easy_profiler/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 2.8.11)
```
Comment (uilianries) [inline recipes/easy_profiler/all/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.1)
```

CXX_STANDARD is CMake 3

File: recipes/easy_profiler/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+from conans import ConanFile, tools, CMake
+import os
+import glob
+
+
+class easy_profilerConan(ConanFile):
+    name = "easy_profiler"
+    description = "Lightweight profiler library for c++"
+    license = "BSD"
```
Comment (uilianries) [inline recipes/easy_profiler/all/conanfile.py:9]:
```suggestion
    license = "MIT"
```

https://github.com/yse/easy_profiler/blob/v2.1.0/LICENSE

File: recipes/easy_profiler/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        os.remove(os.path.join(self.package_folder, "LICENSE.MIT"))
+        os.remove(os.path.join(self.package_folder, "LICENSE.APACHE"))
+
+    def package_info(self):
+        self.cpp_info.libs = ["easy_profiler"]
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["m", "pthread"]
+        self.cpp_info.names["cmake_find_package"] = "easy_profiler"
+        self.cpp_info.names["cmake_find_package_multi"] = "easy_profiler"
```
Comment (uilianries) [inline recipes/easy_profiler/all/conanfile.py:77]:
```suggestion
```

Not required, it uses self.name by default.

File: recipes/easy_profiler/all/test_package/example.cpp [hunk 5]
```diff
@@ -0,0 +1,306 @@
+//#define FULL_DISABLE_PROFILER
```
Comment (uilianries) [inline recipes/easy_profiler/all/test_package/example.cpp:1]: Could you use a smaller example? The idea is checking the package, not exercising the library.

File: recipes/easy_profiler/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+from conans import ConanFile, tools, CMake
+import os
+import glob
```
Comment (uilianries) [inline recipes/easy_profiler/all/conanfile.py:3]:
```suggestion
```

unused

File: recipes/easy_profiler/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class easy_profilerConan(ConanFile):
```
Comment (uilianries) [inline recipes/easy_profiler/all/conanfile.py:6]:
```suggestion
class EasyProfilerConan(ConanFile):
```

https://www.python.org/dev/peps/pep-0008/#id41

File: recipes/easy_profiler/all/test_package/example.cpp (hunk 5 above)
Comment (GavinNL) [reply recipes/easy_profiler/all/test_package/example.cpp:1]: Sure. I just grabbed an example from the original project. Which is usually what I do for the packages I create. It seemed to be okay for the other packages I have created.

File: recipes/easy_profiler/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False]
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True
+    }
```
Comment (uilianries) [inline recipes/easy_profiler/all/conanfile.py:22]:
```suggestion
    }
    short_path = True
```

it requires shot path to pass on Windows build.

**Comment by GavinNL** [comment]
> Some configurations of 'easy_profiler/2.1.0' failed in build 3 (`770bad2c42e3ef80bb1fcd8ce21e978b134419d1`):
> 
> * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  easy_profiler:shared-False**
>   
>   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
>   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
>   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/)

hmm...the short_path didn't work?

**Comment by uilianries** [comment]
> 
> 
> > Some configurations of 'easy_profiler/2.1.0' failed in build 3 (`770bad2c42e3ef80bb1fcd8ce21e978b134419d1`):
> > 
> > * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  easy_profiler:shared-False**
> >   
> >   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
> >   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
> >   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> > * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/3/easy_profiler/2.1.0/)
> 
> hmm...the short_path didn't work?

Sorry, my mistake, the correct name is `short_paths`

**Comment by GavinNL** [comment]
> Some configurations of 'easy_profiler/2.1.0' failed in build 4 (`4e4bcaf721fd77db0490ff51354d96700163e7e5`):
> 
> * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  easy_profiler:shared-False**
>   
>   * Hooks errors detected:
>     
>     * `[HOOK - conan-center.py] post_package(): ERROR: [MS RUNTIME FILES (KB-H021)] The conan-center repository doesn't allow Microsoft Visual Studio runtime files. (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H021) `
>     * `[HOOK - conan-center.py] post_package(): ERROR: [MS RUNTIME FILES (KB-H021)] Found files:`
>   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/4/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
>   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/4/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
>   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/4/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/4/easy_profiler/2.1.0/)

Can anyone suggest the change for this? I don't use windows/MSVC so I don't quite know what needs to be deleted

**Comment by uilianries** [comment]
Hi @GavinNL ! You can read the entire Windows output by the log: https://c3i.jfrog.io/c3i/misc/logs/pr/2441/4/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt (https://github.com/conan-io/conan-center-index/pull/2441#issuecomment-670654045)

Anyway, easy_profiler install:

```
  -- Installing: C:/J/c3ipr/s/PR-2441/4/0_2/10a4cb/1/bin/msvcp140.dll
  -- Installing: C:/J/c3ipr/s/PR-2441/4/0_2/10a4cb/1/bin/vcruntime140.dll
  -- Installing: C:/J/c3ipr/s/PR-2441/4/0_2/10a4cb/1/bin/concrt140.dll
```

Basically you need to remove them, you can use Jasper's idea: https://github.com/conan-io/conan-center-index/blob/dd283c2f1af3eae9136835893e51eca200e1f7ea/recipes/jasper/all/conanfile.py#L75

**Comment by GavinNL** [comment]
> Some configurations of 'easy_profiler/2.1.0' failed in build 6 (`3af2c0b221cdf8d84aa59e4b01c1759c25dc9a5b`):
> 
> * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  easy_profiler:shared-False**
>   
>   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/6/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
>   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/6/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
>   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/6/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/6/easy_profiler/2.1.0/)

I think I need to add a compile-time definition `EASY_OPTION_LIB_STATIC=1` when building a static library on windows. Is there any way to add that through conan without modifying the cmake file?

**Comment by GavinNL** [comment]
> Some configurations of 'easy_profiler/2.1.0' failed in build 7 (`230f1a83dca5db1c22d04ed4bd5ae6818e77317e`):
> 
> * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  easy_profiler:shared-False**
>   
>   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/7/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
>   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/7/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
>   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/7/easy_profiler/2.1.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/2441/7/easy_profiler/2.1.0/)

hmm...did I do something wrong with the way I disabled Windows/Static builds?

File: recipes/easy_profiler/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        # The windows build seems to be giving problems due to certain symbols
+        # not being exported properly for static libraries.
+        if self.options.shared is False and self.settings.os == "Windows":
```
Comment (jgsogo) [inline recipes/easy_profiler/all/conanfile.py:45]:
```suggestion
        if not self.options.shared and self.settings.os == "Windows":
```

The `is` comparator is a bit tricky in Python, we need to revisit it in Conan 2.0: https://github.com/conan-io/conan/issues/3620

File: recipes/easy_profiler/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
+from conans import ConanFile, tools, CMake
```
Comment (jgsogo) [inline recipes/easy_profiler/all/conanfile.py:1]:
```suggestion
from conans import ConanFile, tools, CMake
from conans.errors import ConanInvalidConfiguration
```

File: recipes/easy_profiler/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        # The windows build seems to be giving problems due to certain symbols
+        # not being exported properly for static libraries.
+        if not self.options.shared and self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("Must be built as shared on \
+              Windows")
```
Comment (SpaceIm) [inline recipes/easy_profiler/all/conanfile.py:48]: On Windows, if static, does it fail at build or consume time?

As consume time, seems that `EASY_PROFILER_STATIC` must be defined if static and Windows: https://github.com/yse/easy_profiler/blob/046d9bd40cfd55ca3b1d798b16ad2a5673f66806/easy_profiler_core/include/easy/details/easy_compiler_support.h#L56

Comment (GavinNL) [reply recipes/easy_profiler/all/conanfile.py:48]: I tried that, but it seems there are still some unresolved symbols.

Comment (SpaceIm) [reply recipes/easy_profiler/all/conanfile.py:48]: Ok I tried your branch. Static seems fine, but `psapi` must be added to `cpp_info.system_libs` for windows (unresolved symbols to `GetModuleBaseNameA`).

File: recipes/easy_profiler/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            for dll_file in \
+              glob.glob(os.path.join(self.package_folder, "bin", "*.dll")):
+                if os.path.basename(dll_file).startswith(("concrt", "msvcp",
+                   "vcruntime")):
+                    os.remove(dll_file)
+
+    def package_info(self):
+        self.cpp_info.libs = ["easy_profiler"]
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["m", "pthread"]
```
Comment (SpaceIm) [inline recipes/easy_profiler/all/conanfile.py:88]:
```suggestion
            self.cpp_info.system_libs = ["m", "pthread"]
        elif self.settings.os == "Windows":
            self.cpp_info.system_libs = ["psapi"]
            if not self.options.shared:
                self.cpp_info.defines.append("EASY_PROFILER_STATIC")
```

File: recipes/easy_profiler/all/conanfile.py (hunk 11 above)
Comment (SpaceIm) [reply recipes/easy_profiler/all/conanfile.py:48]:
```suggestion
```

File: recipes/easy_profiler/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
+from conans import ConanFile, tools, CMake
+from conans.errors import ConanInvalidConfiguration
```
Comment (SpaceIm) [inline recipes/easy_profiler/all/conanfile.py:2]:
```suggestion
```

File: recipes/easy_profiler/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    exports_sources = ["CMakeLists.txt", "patches/*"]
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False]
+    }
+    default_options = {
+        "shared": True,
```
Comment (SpaceIm) [inline recipes/easy_profiler/all/conanfile.py:22]:
```suggestion
        "shared": False,
```

File: recipes/easy_profiler/all/conanfile.py (hunk 12 above)
Comment (SpaceIm) [reply recipes/easy_profiler/all/conanfile.py:88]: https://github.com/yse/easy_profiler/blob/a3723894c2d3fba96c758c25f3fa57c96369670d/easy_profiler_core/CMakeLists.txt#L250

File: recipes/easy_profiler/all/conanfile.py (hunk 11 above)
Comment (GavinNL) [reply recipes/easy_profiler/all/conanfile.py:48]: @SpaceIm , could you test it again using ( https://github.com/yse/easy_profiler/blob/develop/sample/main.cpp ) as the test package code. I was using this for the conan test_package and that's when it was failing for me. I then switched to a simpler example as per @uilianries  request.

Comment (SpaceIm) [inline recipes/easy_profiler/all/conanfile.py:48]:
```suggestion
```

Comment (SpaceIm) [reply recipes/easy_profiler/all/conanfile.py:48]: No problem on my computer, it works on Windows 10 if static.

Comment (GavinNL) [reply recipes/easy_profiler/all/conanfile.py:48]: Perfect, add your changes then!

**Pull request merged**